django/tasks/utils.py

```diff
@@ -0,0 +1,71 @@
+import inspect
+import json
+import time
+from functools import wraps
+from traceback import format_exception
+
+from django.utils.module_loading import import_string
+
+
+def is_module_level_function(func):
+    if not inspect.isfunction(func) or inspect.isbuiltin(func):
+        return False
+
+    if "<locals>" in func.__qualname__:
+        return False
+
+    return True
+
+
+def json_normalize(obj):
+    """
+    Round-trip encode object as JSON to normalize types.
+    """
+    return json.loads(json.dumps(obj))
+
+
+def retry(*, retries=3, backoff_delay=0.1):
+    """
+    Retry the given code `retries` times, raising the final error.
+
+    `backoff_delay` can be used to add a delay between attempts.
+    """
+
+    def wrapper(f):
+        @wraps(f)
+        def inner_wrapper(*args, **kwargs):
+            for attempt in range(1, retries + 1):
+                try:
+                    return f(*args, **kwargs)
+                except KeyboardInterrupt:
+                    # Let the user ctrl-C out of the program without a retry
+                    raise
+                except BaseException:
+                    if attempt == retries:
+                        raise
+                    time.sleep(backoff_delay)
+
+        return inner_wrapper
+
+    return wrapper
+
+
+def get_module_path(val):
+    return f"{val.__module__}.{val.__qualname__}"
+
+
+def exception_to_dict(exc):
+    return {
+        "exc_type": get_module_path(type(exc)),
+        "exc_args": json_normalize(exc.args),
```

Comment:
Is there an alternative? Avoiding serialization entirely has the same issues.

Reply:
Stringification works more or less universally.

You cannot roundtrip that, but I believe the roundtripping of the exception to be a feature that should be removed.

Having results be required to be JSON-serializable makes sense because we know what results are.

Exceptions are often downstream of outright mistakes, or other unknown operational errors. Having the errors be caught and recorded is good, but I think in those scenarios we have to try really hard to record the error. Dropping information is, IMO, a worse thing than not being able to deserialize the exception. 

Especially given that people will often (in my experience at least) be looking at the exception by hand, not through automated processes.

Reply:
That's sort of what's here though. The traceback is available as a string, but the exception instance itself is also available if needed for automation reasons.

Reply:
Ah of course, I'd forgotten about the internal arguments :sob:  I think you're right - something much simpler is going to be necessary.

A string traceback, plus just the exception class is probably the best compromise